tools/clang/lib/CodeGen/CGHLSLMS.cpp

```diff
@@ -434,25 +443,167 @@ void CGMSHLSLRuntime::AddHLSLIntrinsicOpcodeToFunction(Function *F,
   m_IntrinsicMap.emplace_back(F,opcode);
 }
 
+static clang::Type::ScalarTypeKind GetScalarElementType(QualType qTy) {
+  if (IsHLSLMatType(qTy))
+    return GetScalarElementType(GetHLSLMatElementType(qTy));
+
+  if (IsHLSLVecType(qTy))
+    return GetScalarElementType(GetHLSLVecElementType(qTy));
+
+  if (qTy->isArrayType())
+    return GetScalarElementType(GetArrayElementType(qTy));
+
+  assert(qTy->isScalarType());
+  return qTy->getScalarTypeKind();
+}
+
+static unsigned GetCompSize(QualType qty) {
+  if (qty->isArrayType())
+    return GetCompSize(GetArrayElementType(qty));
+  if (qty->isScalarType())
+    return 1;
+  assert(IsHLSLVecType(qty));
+  return GetHLSLVecSize(qty);
```

Comment:
`GetScalarElementType` is set up to handle matrices, but this is not. It seems both are ultimately invoked from the same place. I would expect both or neither to expect matrices.

Reply:
good catch!